File: Data Structures & Algorithms/mergeSort/submission-1.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value

from collections import deque

def merge(l1, l2):
    l1 = deque(l1)
    l2 = deque(l2)
    res = []
    while l1 and l2:
        if l1[0].key <= l2[0].key:
            res.append(l1.popleft())
        else:
            res.append(l2.popleft())
    res += l1
    res += l2
    return res


class Solution:
    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        if len(pairs) < 2:
            return pairs
        mid = len(pairs)//2
        left = self.mergeSort(pairs[:mid])
        right = self.mergeSort(pairs[mid:])
        return merge(left, right)

File: Data Structures & Algorithms/sort-an-array/submission-1.go
func merge(left []int, right []int) []int {
    l, r := 0, 0
    res := []int{}
    for l < len(left) && r < len(right) {
        if left[l] < right[r] {
            res = append(res, left[l])
            l++
        } else {
            res = append(res, right[r])
            r++
        }
    }
    for l < len(left) {
        res = append(res, left[l])
        l++
    }
    for r < len(right) {
        res = append(res, right[r])
        r++
    }
    return res
}

func sortArray(nums []int) []int {
    if len(nums) < 2 {
        return nums
    }
    mid := len(nums) / 2
    left := sortArray(nums[:mid])
    right := sortArray(nums[mid:])
    return merge(left, right)
}
